<a href="https://colab.research.google.com/github/jooyoung227/ESAA-14-23-/blob/main/ESAA_jooyoung_YB_week2_study.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#04 데이터 핸들링 - 판다스(Pandas)
판다스의 핵심 개체 Dataframe은 여러 개의 행과 열로 이뤄진 2차원 데이터를 담는 데이터 구조체임

###판다스 시작 - 파일을 DataFrame으로 로딩, 기본 API

In [3]:
import pandas as pd

In [3]:
from google.colab import files
uploaded = files.upload()

import pandas as pd
titanic_df = pd.read_csv('train.csv')

Saving train.csv to train.csv


In [10]:
print('titanic 변수 type:', type(titanic_df))
titanic_df

titanic 변수 type: <class 'pandas.core.frame.DataFrame'>


,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S
...,...,...,...,...,...,...,...,...,...,...,...,...
886,887,0,2,"Montvila, Rev. Juozas",male,27.0,0,0,211536,13.0000,NaN,S
887,888,1,1,"Graham, Miss. Margaret Edith",female,19.0,0,0,112053,30.0000,B42,S
888,889,0,3,"Johnston, Miss. Catherine Helen ""Carrie""",female,NaN,1,2,W./C. 6607,23.4500,NaN,S
889,890,1,1,"Behr, Mr. Karl Howell",male,26.0,0,0,111369,30.0000,C148,C


DataFrame 객체에서 데이터 파일의 첫 줄의 칼럼 문자열이 DataFrame의 칼럼으로 할당됨  
`read_csv()`는 별다른 파라미터 지정이 없으면 파일의 맨 처음 로우를 칼럼명으로 인지하고 칼럼으로 변환  
모든 DataFrame 내의 데이터는 생성되는 순간 고유의 index 값을 가짐

In [11]:
titanic_df.head(3)

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S


In [12]:
print('DataFrame 크기:', titanic_df.shape)

DataFrame 크기: (891, 12)


In [13]:
titanic_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


891 row, 12 column  
2개 칼럼이 float64, 5개 칼럼이 int64, 5개 칼럼이 object 타입

`info()`: 총 데이터 건수와 데이터 타입, Null 건수를 알 수 있음  
`describe()`: 칼럼별 숫자형 데이터 값의 n-percetile 분포도, 평균값, 최댓값, 최솟값을 나타냄, 오직 숫자형 칼럼의 분포도만 조사하며 자동으로 object 타입의 칼럼은 출력에서 제외함  
데이터 분포도를 아는 것은 머신러닝 알고리즘의 성능을 향상시키는 중요한 요소임

In [5]:
titanic_df.describe()

,PassengerId,Survived,Pclass,Age,SibSp,Parch,Fare
count,891.000000,891.000000,891.000000,714.000000,891.000000,891.000000,891.000000
mean,446.000000,0.383838,2.308642,29.699118,0.523008,0.381594,32.204208
std,257.353842,0.486592,0.836071,14.526497,1.102743,0.806057,49.693429
min,1.000000,0.000000,1.000000,0.420000,0.000000,0.000000,0.000000
25%,223.500000,0.000000,2.000000,20.125000,0.000000,0.000000,7.910400
50%,446.000000,0.000000,3.000000,28.000000,0.000000,0.000000,14.454200
75%,668.500000,1.000000,3.000000,38.000000,1.000000,0.000000,31.000000
max,891.000000,1.000000,3.000000,80.000000,8.000000,6.000000,512.329200


DataFrame의 [] 연산자 내부에 칼럼명을 입력하면 Series 형태로 특정 칼럼 데이터 세트가 반환됨. 여기에 `value_counts()`메서드를 호출하면 해당 칼럼값의 유형과 건수를 확인할 수 있음.

In [6]:
value_counts = titanic_df['Pclass'].value_counts()
print(value_counts)

Pclass
3    491
1    216
2    184
Name: count, dtype: int64


In [7]:
titanic_Pclass = titanic_df['Pclass']
print(type(titanic_Pclass))

<class 'pandas.core.series.Series'>


In [9]:
titanic_Pclass.head()

,Pclass
0,3
1,1
2,3
3,1
4,3


단일 칼럼으로 되어있는 series 객체에서 `value_counts()` 메서드를 호출하는 것이 칼럼별 데이터 값의 분포도를 좀 더 명시적으로 파악하기 쉬움

In [10]:
value_counts = titanic_df['Pclass'].value_counts()
print(type(value_counts))
print(value_counts)

<class 'pandas.core.series.Series'>
Pclass
3    491
1    216
2    184
Name: count, dtype: int64


왼쪽이 인덱스 값이고 오른쪽이 데이터 값임  
Pclass 칼럽 값이 3,1,2을 나타내고 있는데 이처럼 인덱스는 단순히 순차 값과 같은 의미없는 식별자만 할당하는 것이 아니라 고유성이 보장된다면 의미 있는 데이터값 할당도 가능함  
`value_counts()`는 칼럼 값별 데이터 건수를 반환하므로 고유 칼럼 값을 식별자로 활용 가능  
인덱스는 DataFrame, series가 생성된 후에도 변경 가능  
인덱스는 숫자형 문자형 모두 가능하나 고유성이 보장되어야 함  

`value_counts()`는 Null 값을 무시하고 결괏값을 내놓기 쉽다는 점을 유의해야함
`value_counts()`는 Null 값을 포함하여 개별 데이터 값의 건수를 계산할지를 dropna 인자로 판단하는 데 dropna의 기본값은 True이며 Null값을 무시하고 개별 데이터 값의 건수를 계산

In [12]:
print('titanic_df 데이터 건수:', titanic_df.shape[0])
print('기본 설정인 dropna=True로 value_counts()')
#value_counts()는 디폴트로 dropna=True이므로 value_counts(dropna=True)와 동일
print(titanic_df['Embarked'].value_counts())
print(titanic_df['Embarked'].value_counts(dropna=False))

titanic_df 데이터 건수: 891
기본 설정인 dropna=True로 value_counts()
Embarked
S    644
C    168
Q     77
Name: count, dtype: int64
Embarked
S      644
C      168
Q       77
NaN      2
Name: count, dtype: int64


###DataFrame과 리스트, 딕셔너리, 넘파이 ndarray 상호 변환
**넘파이 ndarray, 리스트, 딕셔너리를 DataFrame으로 변환하기**
DataFrame은 행과 열을 가지는 2차원 데이터로 2차원 이하의 데이터들만 DataFrame으로 변환할 수 있다

In [13]:
import numpy as np

col_name1=['col1']
list1 = [1, 2, 3]
array1 = np.array(list1)
print('array1 shape:', array1.shape )

#리스트를 이용해 DataFrame 생성
df_list1 = pd.DataFrame(list1, columns=col_name1)
print('1차원 리스트로 만든 DataFrame:\n', df_list1)

#넘파이 ndarray를 이용해 DataFrame 생성
df_array1 = pd.DataFrame(array1, columns=col_name1)
print('1차원 ndarray로 만든 DataFrame:\n', df_array1)

array1 shape: (3,)
1차원 리스트로 만든 DataFrame:
    col1
0     1
1     2
2     3
1차원 ndarray로 만든 DataFrame:
    col1
0     1
1     2
2     3


1차원 형태의 데이터를 기반으로 DataFrame을 생성하므로 칼럼명이 한 개만 필요하다는 사실에 주의

In [14]:
#3개의 칼럼명이 필요함
col_name2 = ['col1', 'col2', 'col3']

#2행X3열 형태의 리스트와 ndarray 생성한 뒤 이를 DataFrame으로 변환
list2 = [[1,2,3],
         [11, 12, 13]]
array2 = np.array(list2)
print('array2 shape:', array2.shape)
df_list2 = pd.DataFrame(list2, columns=col_name2)
print('2차원 리스트로 만든 DataFrame:\n', df_list2)
df_array2 = pd.DataFrame(array2, columns=col_name2)
print('2차원 ndarray로 만든 DataFrame:\n', df_array2)

array2 shape: (2, 3)
2차원 리스트로 만든 DataFrame:
    col1  col2  col3
0     1     2     3
1    11    12    13
2차원 ndarray로 만든 DataFrame:
    col1  col2  col3
0     1     2     3
1    11    12    13


In [15]:
#key는 문자열 칼럼명으로 매핑, value는 리스트형 또는 ndarray 칼럼 데이터로 매핑
dict = {'col1':[1, 11], 'col2':[2, 22], 'col3':[3, 33]}
df_dict = pd.DataFrame(dict)
print('딕셔너리로 만든 DataFrame:\n', df_dict)

딕셔너리로 만든 DataFrame:
    col1  col2  col3
0     1     2     3
1    11    22    33


**DataFrame을 넘파이 ndarray, 리스트, 딕셔너리로 변환하기**

In [16]:
#DataFrame을 ndarray로 변환
array3 = df_dict.values
print('df_dict.values 타입:', type(array3), 'df_dict.values shape:', array3.shape)
print(array3)

df_dict.values 타입: <class 'numpy.ndarray'> df_dict.values shape: (2, 3)
[[ 1  2  3]
 [11 22 33]]


DataFrame 객체의 `to_dict()` 메서드를 호출하고 인자로 list를 입력하면 딕셔너리 값이 리스트형으로 반환됨

In [17]:
#DataFrame을 리스트로 변환
list3 = df_dict.values.tolist()
print('df_dixt.values.tolist() 타입:', type(list3))
print(list3)

#DataFrame을 딕셔너리로 전환
dict3 = df_dict.to_dict('list')
print('\n df_dict.to_dict() 타입:', type(dict3))
print(dict3)

df_dixt.values.tolist() 타입: <class 'list'>
[[1, 2, 3], [11, 22, 33]]

 df_dict.to_dict() 타입: <class 'dict'>
{'col1': [1, 11], 'col2': [2, 22], 'col3': [3, 33]}


###DataFrame의 칼럼 데이터 세트 생성과 수정
DataFrame [] 내에 새로운 칼럼명을 입력하고 값을 할당해주면 쉽게 데이터 세트 생성과 수정이 가능함

In [5]:
titanic_df['Age_0']=0
titanic_df.head(3)

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked,Age_0
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S,0
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C,0
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S,0


In [8]:
#기존 칼럼 series의 데이터를 이용해 새로운 칼럼 series 만들기
titanic_df['Age_by_10'] = titanic_df['Age']*10
titanic_df['Family_No'] = titanic_df['SibSp'] + titanic_df['Parch']+11
titanic_df.head(3)

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked,Age_0,Age_by_10,Family_No
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S,0,220.0,12
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C,0,380.0,12
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S,0,260.0,11


In [9]:
#새롭게 추가한 'Age_by_10'에 일괄적으로 +100
titanic_df['Age_by_10'] = titanic_df['Age_by_10']+100
titanic_df.head(3)

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked,Age_0,Age_by_10,Family_No
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S,0,320.0,12
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C,0,480.0,12
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S,0,360.0,11


#DataFrame 데이터 삭제  
`drop()` 메서드의 원형  
DataFrame.drop(**labels=None, axis=0**, index=None, columns=None, level=None, **inplace=False**, errors='raise')  

axis 0은 로우 방향 축, axis 1은 칼럼 방향 축  
판다스의 DataFrame은 2차원 데이터만 다루므로 axis 0, axis 1로만 axis가 구성  
drop( ) 메서드에 axis=l을 입력하면 칼럼 축 방향으로 드롭을 수행

In [6]:
titanic_drop_df = titanic_df.drop('Age_0', axis=1 )
titanic_drop_df.head(3)

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S


In [7]:
titanic_df.head(3)

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked,Age_0
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S,0
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C,0
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S,0


앞의 예제에서 inplace = False라고 설정했기 때문에 삭제됐다고 생각했던 ‘Age_0’ 칼럼이 여전히 존재함  
 Inplace=False이면 자기 자신의 DataFrame의 데이터는 삭제하지 않
으며, 삭제된 결과 DataFrame을 반환함  
titanic_df DataFrame 객체 변수는 ‘Age_0'  
칼럼이 삭제되지 않았으며 titanic_df.drop('Age_01, axis=l)의 수행 결과로 반환된 titanic_drop_df의 DataFrame 객체 변수만 ‘Age_0’ 칼럼이 삭제됨  
inplace=True로 설정하면 자신의 DataFrame의 데이터를 삭제함  
또한 여러 개의 칼럼을 삭제하고 싶으면 리스트 형태로 삭제하고자 하는 칼럼명을 입력해 labels 파라미터로 입력하면 됨

In [10]:
drop_result = titanic_df.drop(['Age_0', 'Age_by_10', 'Family_No'], axis=1, inplace=True)
print('inplace=True로 drop 후 반환된 값:', drop_result)
titanic_df.head(3)

inplace=True로 drop 후 반환된 값: None


,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S


유의할 점: drop() 시 inplace=True로 설정하면 반환 값이 None이됨  
따라서 inplace=True로 설정한 채로 반환 값을 다시 자신의 DataFrame 객체로 할당 불가

In [11]:
pd.set_option('display.width', 1000)
pd.set_option('display.max_colwidth', 15)
print('#### before axis 0 drop ####')
print(titanic_df.head(3))

titanic_df.drop([0,1,2], axis=0, inplace=True)

print('#### after axis 0 drop ####')
print(titanic_df.head(3))

#### before axis 0 drop ####
   PassengerId  Survived  Pclass            Name     Sex   Age  SibSp  Parch          Ticket     Fare Cabin Embarked
0            1         0       3  Braund, Mr....    male  22.0      1      0       A/5 21171   7.2500   NaN        S
1            2         1       1  Cumings, Mr...  female  38.0      1      0        PC 17599  71.2833   C85        C
2            3         1       3  Heikkinen, ...  female  26.0      0      0  STON/O2. 31...   7.9250   NaN        S
#### after axis 0 drop ####
   PassengerId  Survived  Pclass            Name     Sex   Age  SibSp  Parch  Ticket     Fare Cabin Embarked
3            4         1       1  Futrelle, M...  female  35.0      1      0  113803  53.1000  C123        S
4            5         0       3  Allen, Mr. ...    male  35.0      0      0  373450   8.0500   NaN        S
5            6         0       3  Moran, Mr. ...    male   NaN      0      0  330877   8.4583   NaN        Q


삭제 후 맨 앞 3개 데이터의 인덱스가 3, 4, 5임을 확인할 수 있음

- axis： DataFrame의 로우를 삭제할 때는 axis=0, 칼럼을 삭제할 때는 axis=l으로 설정.
- 원본 DataFrame은 유지하고 드롭된 DataFrame을 새롭게 객체 변수로 받고 싶다면 inplace三False로 설정(디폴트 값이 False임).  
예: titanic_drop_df = titanic_df.drop(' Age_0', axis=1, inplace=False )
- 원본 DataFrame에 드롭된 결과를 적용할 경우에는 inplace=True를 적용.  
예: titanic_df.drop('Age_0', axis=l, inplace=True )
- 원본 DataFrame에서 드롭된 DataFrame을 다시 원본 DataFrame 객체 변수로 할당하면 원본 DataFrame에서 드롭된 결과를 적용할 경우와 같음(단, 기존 원본 DataFrame 객체 변수는 메모리에서 추후 제거됨).  
예: titanic_df = titanic_df.drop(' Age_0', axis=1, inplace=False )


### Index 객체
판다스의 Index 객체는 RDBMS의 PK(Primary Key)와 유사하게 DataFrame, Series의 레코드를 고유하게 식별하는 객체임  
DataFrame, Series에서 Index 객체만 추출하려면 DataFrame.index
또는 Series.index 속성을 통해 가능함

In [12]:
#원본 파일 다시 로딩
titanic_df = pd.read_csv('train.csv')
#Index 객체 추출
indexes = titanic_df.index
#Index 객체를 실제 값 array로 변환
print('Index 객체 arrayr 값:\n', indexes.values)

Index 객체 arrayr 값:
 [  0   1   2   3   4   5   6   7   8   9  10  11  12  13  14  15  16  17
  18  19  20  21  22  23  24  25  26  27  28  29  30  31  32  33  34  35
  36  37  38  39  40  41  42  43  44  45  46  47  48  49  50  51  52  53
  54  55  56  57  58  59  60  61  62  63  64  65  66  67  68  69  70  71
  72  73  74  75  76  77  78  79  80  81  82  83  84  85  86  87  88  89
  90  91  92  93  94  95  96  97  98  99 100 101 102 103 104 105 106 107
 108 109 110 111 112 113 114 115 116 117 118 119 120 121 122 123 124 125
 126 127 128 129 130 131 132 133 134 135 136 137 138 139 140 141 142 143
 144 145 146 147 148 149 150 151 152 153 154 155 156 157 158 159 160 161
 162 163 164 165 166 167 168 169 170 171 172 173 174 175 176 177 178 179
 180 181 182 183 184 185 186 187 188 189 190 191 192 193 194 195 196 197
 198 199 200 201 202 203 204 205 206 207 208 209 210 211 212 213 214 215
 216 217 218 219 220 221 222 223 224 225 226 227 228 229 230 231 232 233
 234 235 236 237 238 239 240 24

In [13]:
print(type(indexes.values))
print(indexes.values.shape)
print(indexes[:5].values)
print(indexes.values[:5])
print(indexes[6])

<class 'numpy.ndarray'>
(891,)
[0 1 2 3 4]
[0 1 2 3 4]
6


Index 객체는 식별성 데이터를 1차원 array로 가지고 있음  
ndarray와 유사하게 단일 값 반환 및 슬라이싱도 가능  
한 번 만들어진 DataFrame 및 series의 Index 객체는 함부로 변경 불가

In [14]:
indexes[0]=5

TypeError: Index does not support mutable operations

In [18]:
series_fair = titanic_df['Fare']
print('Fair Series max 값:', series_fair.max())
print('Fair Series sum 값:', series_fair.sum())
print('sum() Fair Series：',sum(series_fair))
print('Fair Series + 3:\n',(series_fair + 3).head(3))

Fair Series max 값: 512.3292
Fair Series sum 값: 28693.9493
sum() Fair Series： 28693.9493
Fair Series + 3:
 0    10.2500
1    74.2833
2    10.9250
Name: Fare, dtype: float64


DataFrame 및 Series에 reset_index() 메서드를 수행하면 새롭게 인덱스를 연속 숫자 형으로 할당하며 기존 인덱스는 ‘index’라는 새로운 칼럼 명으로 추가함

In [21]:
titanic_reset_df = titanic_df.reset_index(inplace=False)
titanic_reset_df.head(3)

,index,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,0,1,0,3,"Braund, Mr....",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,1,2,1,1,"Cumings, Mr...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,2,3,1,3,"Heikkinen, ...",female,26.0,0,0,STON/O2. 31...,7.9250,NaN,S


`reset_index()`는 인덱스가 연속된 int 숫자형 데이터가 아닐 경우에 다시 이를 연속 int 숫자형 데이터로 만들 때 주로 사용  
유의할점: Series에 reset_index()를 적용하면 Series가 아닌 DataFrame이 반환됨  
reset_index( )의 parameter 중 drop=Tme로 설정하면 기존 인덱스는 새로운 칼럼으로 추가되지 않고 삭제(drop)되며 새로운 칼럼이 추가되지 않으므로 series로 유지됨

In [22]:
print( '### before reset_index ###')
value_counts = titanic_df['Pclass'].value_counts()
print(value_counts)
print('value_counts 객체 변수 타입:', type(value_counts))
new_value_counts = value_counts.reset_index(inplace=False)
print( '### After reset_index ###')
print(new_value_counts)
print('new_value_counts 객체 변수 타입:', type(new_value_counts))

### before reset_index ###
Pclass
3    491
1    216
2    184
Name: count, dtype: int64
value_counts 객체 변수 타입: <class 'pandas.core.series.Series'>
### After reset_index ###
   Pclass  count
0       3    491
1       1    216
2       2    184
new_value_counts 객체 변수 타입: <class 'pandas.core.frame.DataFrame'>


###데이터 셀렉션 및 필터링
**DataFrame의 [ ] 연산자**  
 DataFrame 바로 뒤에 있는 ‘[ ]’ 안에 들어갈 수 있는 것은 칼럼명 문자(또는 칼럼명의 리스트 객체), 또는 인덱스로 변환 가능한 표현식  
 현재 수준에서는 DataFrame 뒤에 있는 [ ]는 칼럼만 지정할 수 있는 ‘칼럼 지정
연산자’로 이해하는 게 혼돈을 막는 가장 좋은 방법  

넘파이에서 [ ] 연산자는 행의 위치, 열의 위치, 슬라이싱 범위 등을 지정해 데이터를 가져올 수 있

In [23]:
print('단일 칼럼 데이터 추출:\n', titanic_df[ 'Pclass'].head(3))
print('\n여러 칼럼의 데이터 추출:\n', titanic_df[ ['Survived', 'Pclass'] ].head(3))
print('[ ] 안에 숫자 index는 KeyError 오류 발생:\n', titanic_df[0])

단일 칼럼 데이터 추출:
 0    3
1    1
2    3
Name: Pclass, dtype: int64

여러 칼럼의 데이터 추출:
    Survived  Pclass
0         0       3
1         1       1
2         1       3


KeyError: 0

In [25]:
titanic_df[ 0:2 ]

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr....",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mr...",female,38.0,1,0,PC 17599,71.2833,C85,C


In [26]:
#불린 인덱싱 표현도 가능
titanic_df[titanic_df['Pclass']==3].head(3)

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr....",male,22.0,1,0,A/5 21171,7.250,NaN,S
2,3,1,3,"Heikkinen, ...",female,26.0,0,0,STON/O2. 31...,7.925,NaN,S
4,5,0,3,"Allen, Mr. ...",male,35.0,0,0,373450,8.050,NaN,S


- DataFrame 바로 뒤의 [ ] 연산자는 넘파이의 [ ]나 Series의 [ ]와 다름
- DataFrame 바로 뒤의 [ ] 내 입력값은 칼럼명(또는 칼럼의 리스트)을 지정해 칼럼 지정 연산에 사용하거나 불린 인덱스 용도로만 사용해야 함.
- DataFrame[ 0:2 ]와 같은 슬라이싱 연산으로 데이터를 추출하는 방법은 사용하지 않는 게 좋음

**DataFrame iloc[ ] 연산자**  
iloc[]는 위치(Location) 기반 인덱싱 방식으로 동작하며 loc[ ]는 명칭(Label)기
반 인덱싱 방식으로 동작  
iloc[ ]는 위치 기반 인덱싱만 허용하기 때문에 행과 열의 좌표 위치에 해당하는 값으로 정숫값 또는 정
수형의 슬라이싱, 팬시 리스트 값을 입력해줘야 함

In [27]:
data = {'Name': ['Chulmin', 'Eunkyung','Jinwoong', 'Soobeom'],
        'Year': [2011, 2016, 2015, 2015],
        'Gender': ['Male', 'Female', 'Male', 'Male']
        }
data_df = pd.DataFrame(data, index=['one','two','three','four'])
data_df

,Name,Year,Gender
one,Chulmin,2011,Male
two,Eunkyung,2016,Female
three,Jinwoong,2015,Male
four,Soobeom,2015,Male


In [30]:
data_df.iloc[0, 0]

'Chulmin'

In [31]:
# 아래 코드는 오류를 발생시킵니다.
data_df.iloc[0, 'Name']

ValueError: Location based indexing can only have [integer, integer slice (START point is INCLUDED, END point is EXCLUDED), listlike of integers, boolean array] types

In [33]:
# 아래 코드는 오류를 발생합니다.
data_df.iloc['one', 0]

ValueError: Location based indexing can only have [integer, integer slice (START point is INCLUDED, END point is EXCLUDED), listlike of integers, boolean array] types

`data_df.iloc[1,0]`: 두번째 행의 첫번째 열 위치에 있는 단일 값 반환  
`data_df.iloc[2,l]`: 세번째 행의 두번째 열 위치에 있는 단일 값 반환  
`data_df.iloc[0：2, [0,1]]`: 0：2 슬라이싱 범위의 첫번째에서 두번째 행과 첫번째, 두번째 열에 해당하는 DataFrame 반환  
`data_df.iloc[0：2, 0：3]`: 0：2 슬라이싱 범위의 첫번째에서 두번째 행의 0：3 슬라이싱 범위의 첫번째부터 세번째 열 범위에 해당하는 DataFrame 반환  
`data_df.iloc[:]`: 전체 DataFrame 반환  
`data_df.iloc[:, ：]`: 전체 DataFrame 반환

'-1': 맨 마지막 데이터 값  
iloc[：, -1] : 맨 마지막 칼럼의 값, 타깃 값  
 iloc[:, ：-1]: 하게 되면 처음부터 맨 마지막 칼럼을 제외한 모든 칼럼의 값, 피처값  

 `iloc[]`은 슬라이싱과 팬시 인덱싱은 제공하지만 불린 인덱싱은 제공하지 않음

In [36]:
print("\n 맨 마지막 칼럼 데이터 [:, -1 ] \n", data_df.iloc[:, -1])
print("\n 맨 마지막 칼럼을 제외한 모든 데이터 [:,：-1] \n", data_df.iloc[: , :-1])


 맨 마지막 칼럼 데이터 [:, -1 ] 
 one        Male
two      Female
three      Male
four       Male
Name: Gender, dtype: object

 맨 마지막 칼럼을 제외한 모든 데이터 [:,：-1] 
            Name  Year
one     Chulmin  2011
two    Eunkyung  2016
three  Jinwoong  2015
four    Soobeom  2015


**DataFrame loc[] 연산자**  
`loc[]` 명칭(Label) 기반으로 데이터를 추출

In [37]:
data_df.loc['one', 'Name']

'Chulmin'

In [38]:
# 다음 코드는 오류를 발생시킵니다.
data_df.loc[0, 'Name']

KeyError: 0

loc[ ]에 슬라이싱 기호를 적용하면 종료 값-1이 아니라 종료 값까지
포함하는 것을 의미

In [40]:
print('위치기반 iloc slicing\n', data_df.iloc[0:1, 0],'\n')
print('명칭기반 loc slicing\n', data_df.loc['one' :'two', 'Name'])

위치기반 iloc slicing
 one    Chulmin
Name: Name, dtype: object 

명칭기반 loc slicing
 one     Chulmin
two    Eunkyung
Name: Name, dtype: object


`data_df.loc['three', 'Name']`: 인덱스 값 three인 행의 Name 칼럼의 단일값 반환
`data_df,loc['one'：'two',['Name', 'Year']]`: 인덱스 값 one부터 two까지 행의 Name과 Year 칼럼에 해당하는 DataFrame 반환  
`data_df.loc['one':'three','Name':'Gender']`: 인덱스 값 one부터 three까지 행의 Name부터 Gender 칼럼까지의 DataFrame 반환  
`data_df.loc[:]`: 모든 데이터 값  
`data_df.loc[data_df.Year >= 2014]`: iloc[ ]와 다르게 loc[ ]는 불린 인덱싱이 가능. Year 칼럼의 값이 2014 이상인 모든 데이터를 불린 인덱싱으로 추출

1. 개별 또는 여러 칼럼 값 전체를 추출하고자 한다면 iloc[ ]나 loc[ ]를 사용하지 않고 data_df[‘Name’]과 같이 DataFramef칼럼명]만으로 충분합니다. 하지만 행과 열을 함께 사용하여 데이터를 추출해야 한다면 iloc[ ]나 loc[]를 사용해야 합니다.
2. iloc[ ]와 loc[ ]를 이해하기 위해서는 명칭 기반 인덱싱과 위치 기반 인덱싱의 차이를 먼저 이해해야 합니다.
DataFrame의 인덱스나 칼럼명으로 데이터에 접근하는 것은 명칭 기반 인덱싱입니다. 0부터 시작하는 행, 열의 위
치 좌표에만 의존하는 것이 위치 기반 인덱싱입니다.
3. iloc[｝는 위치 기반 인덱싱만 가능합니다. 따라서 행과 열 위치 값으로 정수형 값을 지정해 원하는 데이터를 반환합
니다.
4. Ioc[ ]는 명칭 기반 인덱싱만 가능합니다. 따라서 행 위치에 DataFrame 인덱스가 오며, 열 위치에는 칼럼명을 지정
해 원하는 데이터를 반환합니다.
5. 명칭 기반 인덱싱에서 슬라이싱을‘시작점:종료점’으로 지정할 때 시작점에서 종료점을 포함한 위치에 있는 데이터를 반환합니다.

**불린 인덱싱**  


In [42]:
titanic_boolean = titanic_df[titanic_df['Age']> 60]
print(type(titanic_boolean))
titanic_boolean

<class 'pandas.core.frame.DataFrame'>


,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
33,34,0,2,"Wheadon, Mr...",male,66.0,0,0,C.A. 24579,10.5000,NaN,S
54,55,0,1,"Ostby, Mr. ...",male,65.0,0,1,113509,61.9792,B30,C
96,97,0,1,Goldschmidt...,male,71.0,0,0,PC 17754,34.6542,A5,C
116,117,0,3,"Connors, Mr...",male,70.5,0,0,370369,7.7500,NaN,Q
170,171,0,1,Van der hoe...,male,61.0,0,0,111240,33.5000,B19,S
252,253,0,1,"Stead, Mr. ...",male,62.0,0,0,113514,26.5500,C87,S
275,276,1,1,"Andrews, Mi...",female,63.0,1,0,13502,77.9583,D7,S
280,281,0,3,"Duane, Mr. ...",male,65.0,0,0,336439,7.7500,NaN,Q
326,327,0,3,"Nysveen, Mr...",male,61.0,0,0,345364,6.2375,NaN,S
438,439,0,1,"Fortune, Mr...",male,64.0,1,4,19950,263.0000,C23 C25 C27,S


In [43]:
titanic_df[titanic_df['Age'] > 60][['Name', 'Age' ]].head(3)

,Name,Age
33,"Wheadon, Mr...",66.0
54,"Ostby, Mr. ...",65.0
96,Goldschmidt...,71.0


In [44]:
titanic_df.loc[titanic_df['Age'] > 60, ['Name', 'Age']].head(3)

,Name,Age
33,"Wheadon, Mr...",66.0
54,"Ostby, Mr. ...",65.0
96,Goldschmidt...,71.0


1. and 조건일 때는 &
2. or 조건일 때는 |
3. Not 조건일 때는 ~

In [49]:
titanic_df[ (titanic_df['Age'] > 60) & (titanic_df['Pclass']==1)& (titanic_df['Sex']=='female')]

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
275,276,1,1,"Andrews, Mi...",female,63.0,1,0,13502,77.9583,D7,S
829,830,1,1,"Stone, Mrs....",female,62.0,0,0,113572,80.0000,B28,NaN


In [51]:
cond1 = titanic_df['Age'] > 60
cond2 = titanic_df ['Pclass']==1
cond3 = titanic_df[ 'Sex' ]=='female'
titanic_df[ cond1 & cond2 & cond3]

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
275,276,1,1,"Andrews, Mi...",female,63.0,1,0,13502,77.9583,D7,S
829,830,1,1,"Stone, Mrs....",female,62.0,0,0,113572,80.0000,B28,NaN


###정렬, Aggregation 흗h수% GroupBy 적용
**DataFrame, Series의 정렬 - sort_values()**  
ascending= True로 설정하면 오름차순으로 정렬, ascending듸False로 설정하면 내림차순으로 정렬  
inplace=False로 설정하면 sort_values( )를 호출한 DataFrame은 그대로 유지하며 정렬된 DataFrame을 결과로 반환함. inplace=True로 설정하면
호출한 DataFrame의 정렬 결과를 그대로 적용

In [52]:
titanic_sorted = titanic_df.sort_values(by=['Name'])
titanic_sorted.head(3)

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
845,846,0,3,"Abbing, Mr....",male,42.0,0,0,C.A. 5547,7.55,NaN,S
746,747,0,3,"Abbott, Mr....",male,16.0,1,1,C.A. 2673,20.25,NaN,S
279,280,1,3,"Abbott, Mrs...",female,35.0,1,1,C.A. 2673,20.25,NaN,S


In [55]:
titanic_sorted = titanic_df.sort_values(by=['Pclass', 'Name'], ascending=False)
titanic_sorted.head(3)

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
868,869,0,3,van Melkebe...,male,NaN,0,0,345777,9.5,NaN,S
153,154,0,3,van Billiar...,male,40.5,0,2,A/5. 851,14.5,NaN,S
282,283,0,3,de Pelsmaek...,male,16.0,0,0,345778,9.5,NaN,S


**Aggregation 함수 적용**  
- min( ), max( ), sum( ), count( )  
count()는 Null 값을 반영하지 않은 결과를 반환합니다. 때문에 Null 값이 있는 Age, Cabin, Embarked 칼럼은 count()의 결과 값이 다름

In [56]:
titanic_df.count()

,0
PassengerId,891
Survived,891
Pclass,891
Name,891
Sex,891
Age,714
SibSp,891
Parch,891
Ticket,891
Fare,891


In [57]:
titanic_df[['Age', 'Fare']].mean()

,0
Age,29.699118
Fare,32.204208


**groupby() 적용**  
입력 파라미터 by에 칼럼을 입력하면 대상 칼럼으로 groupby됨

In [58]:
titanic_groupby = titanic_df.groupby(by='Pclass')
print(type(titanic_groupby))

<class 'pandas.core.groupby.generic.DataFrameGroupBy'>


SQL의 group by와 다르게, DataFrame에 groupby()를 호출해 반환된 결과에 aggregation 함수를 호출하면 groupby() 대상 칼럼을 제외한 모든 칼럼에 해당 aggregation 함수를 적용함

In [59]:
titanic_groupby = titanic_df.groupby('Pclass').count()
titanic_groupby

,PassengerId,Survived,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
Pclass,,,,,,,,,,,
1,216,216,216,216,186,216,216,216,216,176,214
2,184,184,184,184,173,184,184,184,184,16,184
3,491,491,491,491,355,491,491,491,491,12,491


In [61]:
titanic_groupby = titanic_df.groupby('Pclass')[['PassengerId', 'Survived']].count()
titanic_groupby

,PassengerId,Survived
Pclass,,
1,216,216
2,184,184
3,491,491


DataFrame groupby()는 적용하려는 여러 개의 aggregation 함수명을 DataFrameGroupBy 객체의 agg( ) 내에 인자로 입력해서 사용함

In [62]:
titanic_df.groupby('Pclass')['Age'].agg([max, min])

/tmp/ipykernel_536/1565262784.py:1: FutureWarning: The provided callable <built-in function max> is currently using SeriesGroupBy.max. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "max" instead.
  titanic_df.groupby('Pclass')['Age'].agg([max, min])
/tmp/ipykernel_536/1565262784.py:1: FutureWarning: The provided callable <built-in function min> is currently using SeriesGroupBy.min. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "min" instead.
  titanic_df.groupby('Pclass')['Age'].agg([max, min])


,max,min
Pclass,,
1,80.0,0.92
2,70.0,0.67
3,74.0,0.42


In [65]:
agg_format={'Age':'max', 'SibSp':'sum', 'Fare':'mean'}
titanic_df.groupby('Pclass').agg(agg_format)

,Age,SibSp,Fare
Pclass,,,
1,80.0,90,84.154687
2,70.0,74,20.662183
3,74.0,302,13.675550


###결손 데이터(Null) 처리하기
NaN 여부를 확인하는 API: `isna( )`  
NaN 값을 다른 값으로 대체하는 API: `fillna()`  

**isna()로 결손 데이터여부 확인**  
DataFrame에 isna()를 수행하면 모든 칼럼의 값 이 NaN인지 아닌지를 True나 False로 알려

In [66]:
titanic_df.isna().head(3)

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,False,False,False,False,False,False,False,False,False,False,True,False
1,False,False,False,False,False,False,False,False,False,False,False,False
2,False,False,False,False,False,False,False,False,False,False,True,False


In [67]:
titanic_df.isna( ).sum( )

,0
PassengerId,0
Survived,0
Pclass,0
Name,0
Sex,0
Age,177
SibSp,0
Parch,0
Ticket,0
Fare,0


**fillna()로 결손 데이터 대체하기**  

In [68]:
titanic_df['Cabin'] = titanic_df['Cabin'].fillna('C000')
titanic_df.head(3)

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr....",male,22.0,1,0,A/5 21171,7.2500,C000,S
1,2,1,1,"Cumings, Mr...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, ...",female,26.0,0,0,STON/O2. 31...,7.9250,C000,S


In [73]:
titanic_df['Age'] = titanic_df['Age'].fillna(titanic_df['Age'].mean())
titanic_df['Embarked'] = titanic_df['Embarked'].fillna('S')
titanic_df.isna().sum()

,0
PassengerId,0
Survived,0
Pclass,0
Name,0
Sex,0
Age,0
SibSp,0
Parch,0
Ticket,0
Fare,0


###apply lambda 식으로 데이터 가공
판다스는 apply 함수에 lambda 식을 결합해 DataFrame이나 Series의 레코드별로 데이터를 가공하는 기능을 제

In [74]:
def get_square(a):
  return a**2
print('3의 제곱은:', get_square(3))

3의 제곱은: 9


In [77]:
lambda_square = lambda x :x ** 2
print('3의 제곱은:', lambda_square(3))

3의 제곱은: 9


In [78]:
a=[1, 2, 3]
squares = map(lambda x : x**2, a)
list(squares)

[1, 4, 9]

In [81]:
titanic_df['Name_len']= titanic_df['Name'].apply(lambda x:len(x))
titanic_df[['Name', 'Name_len']].head(3)

,Name,Name_len
0,"Braund, Mr....",23
1,"Cumings, Mr...",51
2,"Heikkinen, ...",22


In [83]:
titanic_df['Child_Adult'] = titanic_df['Age'].apply(lambda x : 'Child' if x <=15 else 'Adult' )
titanic_df[['Age', 'Child_Adult']].head(8)

,Age,Child_Adult
0,22.000000,Adult
1,38.000000,Adult
2,26.000000,Adult
3,35.000000,Adult
4,35.000000,Adult
5,29.699118,Adult
6,54.000000,Adult
7,2.000000,Child


lambda 식‘：’ 기호의 오른편에 반환 값이 있어야 하기 때문에 if 절의 경우 if 식보다 반환 값을 먼저 기술  
따라서 lambda x :if x〈=15 'Child' else 'Adult' ㄴㄴ  
lambda x:'Child' if x〈=15 else 'Adult' ㅇㅇ  
else의 경우 else 식이 먼저 나오고 반환 값이 나중에 옴  
else if는 지원 X -> else if를 이용 하기 위해서는 else 절을 ()로 내포해 () 내에서 다시 if else 적용해 사용

In [88]:
titanic_df['Age_cat'] = titanic_df['Age'].apply(lambda x : 'Child' if x<=15 else ('Adult' if x <= 60 else 'Elderly'))
titanic_df['Age_cat'].value_counts()

,count
Age_cat,
Adult,786
Child,83
Elderly,22


In [89]:
# 나이에 따라 세분화된 분류를 수행하는 함수 생성.
def get_category(age):
  cat = ''
  if age <= 5: cat = 'Baby'
  elif age <= 12: cat = 'Child'
  elif age <= 18: cat = 'Teenager'
  elif age <= 25: cat = 'Student'
  elif age <= 35: cat = 'Young Adult'
  elif age <= 60: cat = 'Adult'
  else : cat = 'Elderly'
  return

# lambda 식에 위에서 생성한 get_category( ) 함수를 반환값으로 지정.
# get_category(X)는 입력값으로 'Age' 칼럼 값을 받아서 해당하는 cat 반환
titanic_df['Age_cat'] = titanic_df['Age'].apply(lambda x : get_category(x))
titanic_df[['Age', 'Age_cat']].head()

,Age,Age_cat
0,22.0,None
1,38.0,None
2,26.0,None
3,35.0,None
4,35.0,None


#05 정리
오늘 배운거 중요하다!